rfcs/text/0017_url_service.md

```diff
@@ -0,0 +1,462 @@
+1. Short URLs store only the URL of the destination page. However, the
+   destination page might have other state which affects the display of the page
+   but is not present in the URL. Once the short URL is used to navigate to that
+   page, any state that is kept only in memory is lost.
+   1. __Could do:__ We will not support storing additional, non-URL state in
+      short URLs in the initial implementation. However, it will be possible to
+      store addition non-URL state with each short URL. That state would be
+      provided to a Kibana app once user first navigates to that app using a
+      short URL. (This non-URL state might also need migration support.)
+1. Short URLs are not automatically deleted when the target (say dashboard) is
+   deleted. (#10450)
+   1. __Could do:__ The URL Service will not provide such feature. Though the
+      short URLs will keep track of saved object references used in the params
+      to generate a short URL. Maybe those saved references could somehow be
+      used in the future to provide such facility.
+      
+      Currently, there are two possible avenues for deleting a short URL when
+      the underlying dashboard is deleted:
+
+      1. The Dashboard app could keep track of short URLs it generates for each
+         dashboard. Once a dashboard is deleted, the Dashboard app also
+         deletes all short URLs associated with that dashboard.
+      1. Saved Objects Service could implement *cascading deletes*. Once a saved
+         object is deleted, the associated saved objects are also deleted
+         (#71453).
+1. Add additional metadata to each short URL.
+   1. __Could do:__ Each short URL already keeps a counter of how often it was
+      resolved, we could also keep track of a timestamp when it was last
+      resolved, and have an ability for users to give a title to each short URL.
+1. Short URLs don't have a management UI.
+   1. __Will NOT do:__ We will not create a dedicated UI for managing short
+      URLs. We could improve how short URLs saved objects are presented in saved
+      object management UI.
+1. Short URLs can't be created by read-only users (#18006).
+   1. __Will NOT do:__ Currently short URLs are stored as saved objects of type
+      `url`, we would like to keep it that way and benefit from saved object
+      facilities like references, migrations, authorization etc.. The consensus
+      is that we will not allow anonymous users to create short URLs. We want to
+      continue using saved object for short URLs going forward and not
+      compromise on their security model.
+
+
+## Limitations of the URL Generator Service
+
+We have identified the following limitations in the current implementation of
+the URL Generator Service:
+
+1. URL generators are available only on the client-side. There is no way to use
+   them together with short URLs.
+   1. __Will do:__ We will implement URL generators also on the server-side
```

Comment:
Idea is to change the current RFC so that URL generators return `KibanaLocation`, where:

```ts
interface KibanaLocation {
  app: string; // app ID
  path: string;  // path in app
  state: object; // object to pass using navigateWithState()
}
```

so basically it is a relative URL, and we solve separately how it is converted to an absolute URL.